bokehjs/src/coffee/models/tools/util.ts

```diff
@@ -0,0 +1,33 @@
+import {Renderer} from "../renderers/renderer"
+import {GlyphRenderer} from "../renderers/glyph_renderer"
+import {GraphRenderer} from "../renderers/graph_renderer"
+
+import {includes} from "core/util/array"
+
+export type DataRenderer = GlyphRenderer | GraphRenderer
+
+export type RendererSpec = DataRenderer[] | "auto" | null
+
+export function compute_renderers(renderers: RendererSpec, all_renderers: Renderer[], names: string[]): DataRenderer[] {
+    if (renderers == null) {
+      return []
+    }
+
+    if (renderers.length == 0) {
+      return []
+    }
+
+    let result = renderers as DataRenderer[]
```

Comment:
You can't just assume `renderers` are `DataRenderer`s. The original code correctly filtered `all_renderers` to assert that.

Reply:
In fact the problem starts with`renderers = Either(Auto, List(Instance(Renderer)))` and not using `DataRenderer` (doesn't exist at this point, but I have intended for a while to add one to clarify things).

Reply:
The code immediately below does the filtering (and is identical to the previous code). If you have a concrete suggestion, I am happy to implement it.

Reply:
Ah yes, I see it now. This is even worse that I initially viewed it. This is how the code should look like:
```
  if (renderers == null)
      return []

  let result: DataRenderer[]
  if (renderers == 'auto') {
    result = all_renderers.filter((r): r is DataRenderer => {
      return r instanceof GlyphRenderer || r instanceof GraphRenderer
    })
  } else
    result = renderers

  if (names.length > 0)
    result = result.filter((r) => includes(names, r.name))

  return result
```

Reply:
Also, with the original code:
```ts
   if (renderers.length == 0) {
      return []
    }
```
without proper type inspection, you are effectively checking length of a string when `"auto"`. This isn't incorrect, but definitively strange. TypeScript doesn't help here, because `lenght` is a valid member on string, even if it's a literal string. However, as with tuples, this is likely to get improved and length with be fixed (to 4 in this case), so this code may start failing compilation at some point.

Reply:
I've updated to start `result` as an empty list and push appropriate things to it.